File: collector-bitmovin-player/src/main/java/com/bitmovin/analytics/bitmovin/player/BitmovinPlayerCollector.java
package com.bitmovin.analytics.bitmovin.player;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import com.bitmovin.analytics.BitmovinAnalytics;
import com.bitmovin.analytics.BitmovinAnalyticsConfig;
import com.bitmovin.analytics.Collector;
import com.bitmovin.analytics.DefaultCollector;
import com.bitmovin.analytics.adapters.PlayerAdapter;
import com.bitmovin.analytics.bitmovin.player.features.BitmovinFeatureFactory;
import com.bitmovin.analytics.config.SourceMetadata;
import com.bitmovin.analytics.features.FeatureFactory;
import com.bitmovin.player.api.Player;
import com.bitmovin.player.api.source.Source;
import java.util.HashMap;
import org.jetbrains.annotations.NotNull;

public class BitmovinPlayerCollector extends DefaultCollector<Player> implements Collector<Player> {
    private HashMap<Source, SourceMetadata> sourceMetadataMap = new HashMap<>();

    /**
     * Bitmovin Analytics
     *
     * @param bitmovinAnalyticsConfig {@link BitmovinAnalyticsConfig}
     * @param context {@link Context}
     */
    public BitmovinPlayerCollector(
            BitmovinAnalyticsConfig bitmovinAnalyticsConfig, Context context) {
        super(Companion.createAnalytics(bitmovinAnalyticsConfig, context, getUserAgent(context)));
    }

    /**
     * Bitmovin Analytics
     *
     * @param bitmovinAnalyticsConfig {@link BitmovinAnalyticsConfig}
     * @deprecated Please use {@link #BitmovinPlayerCollector(BitmovinAnalyticsConfig, Context)} and
     *     pass {@link Context} separately.
     */
    @Deprecated
    public BitmovinPlayerCollector(BitmovinAnalyticsConfig bitmovinAnalyticsConfig) {
        this(bitmovinAnalyticsConfig, bitmovinAnalyticsConfig.getContext());
    }

    @NotNull
    @Override
    protected PlayerAdapter createAdapter(Player player, @NotNull BitmovinAnalytics analytics) {
        FeatureFactory featureFactory = new BitmovinFeatureFactory(analytics, player);
        return new BitmovinSdkAdapter(
                player,
                analytics.getConfig(),
                analytics.getPlayerStateMachine(),
                featureFactory,
                sourceMetadataMap);
    }

    public void addSourceMetadata(Source playerSource, SourceMetadata sourceMetadata) {
        sourceMetadataMap.put(playerSource, sourceMetadata);
    }

    @NotNull
    private static String getUserAgent(Context context) {
        ApplicationInfo applicationInfo = context.getApplicationInfo();
        int stringId = applicationInfo.labelRes;
        String applicationName = "Unknown";
        if (stringId == 0 && applicationInfo.nonLocalizedLabel != null) {
            applicationInfo.nonLocalizedLabel.toString();
        }
        String versionName;
        try {
            String packageName = context.getPackageName();
            PackageInfo info = context.getPackageManager().getPackageInfo(packageName, 0);
            versionName = info.versionName;
        } catch (PackageManager.NameNotFoundException var5) {
            versionName = "?";
        }

        return applicationName
                + "/"
                + versionName
                + " (Linux;Android "
                + Build.VERSION.RELEASE
                + ") "
                + "BitmovinPlayer/"
                + BitmovinUtil.getPlayerVersion();
    }
}

File: collector-bitmovin-player/src/main/java/com/bitmovin/analytics/bitmovin/player/BitmovinPlayerCollector.kt
package com.bitmovin.analytics.bitmovin.player

import android.content.Context
import android.os.Handler
import com.bitmovin.analytics.BitmovinAnalytics
import com.bitmovin.analytics.BitmovinAnalyticsConfig
import com.bitmovin.analytics.DefaultCollector
import com.bitmovin.analytics.adapters.PlayerAdapter
import com.bitmovin.analytics.api.AnalyticsConfig
import com.bitmovin.analytics.api.CustomData
import com.bitmovin.analytics.api.SourceMetadata
import com.bitmovin.analytics.bitmovin.player.api.IBitmovinPlayerCollector
import com.bitmovin.analytics.bitmovin.player.features.BitmovinFeatureFactory
import com.bitmovin.analytics.bitmovin.player.player.BitmovinPlayerContext
import com.bitmovin.analytics.bitmovin.player.player.PlaybackQualityProvider
import com.bitmovin.analytics.bitmovin.player.player.PlayerLicenseProvider
import com.bitmovin.analytics.data.DeviceInformationProvider
import com.bitmovin.analytics.data.EventDataFactory
import com.bitmovin.analytics.features.FeatureFactory
import com.bitmovin.analytics.stateMachines.PlayerStateMachine
import com.bitmovin.analytics.utils.ApiV3Utils
import com.bitmovin.analytics.utils.SystemInformationProvider
import com.bitmovin.analytics.utils.UserAgentProvider
import com.bitmovin.analytics.utils.Util
import com.bitmovin.player.api.Player
import com.bitmovin.player.api.source.Source

@Deprecated(
    "Please use {@link IBitmovinCollector.Factory instead} instead.",
    replaceWith = ReplaceWith(
        expression = "IBitmovinPlayerCollector.Factory.create(context, analyticsConfig)",
        imports = ["com.bitmovin.analytics.bitmovin.player.api.IBitmovinPlayerCollector"],
    ),
)
class BitmovinPlayerCollector(analyticsConfig: AnalyticsConfig, context: Context) :
    DefaultCollector<Player>(analyticsConfig, context.applicationContext),
    IBitmovinPlayerCollector {
    private val deferredLicenseManager = DeferredLicenseRelay(analyticsConfig.licenseKey)

    override val analytics: BitmovinAnalytics by lazy {
        BitmovinAnalytics(
            config = analyticsConfig,
            context = context,
            licenseKeyProvider = deferredLicenseManager.licenseKeyProvider,
        )
    }

    override fun attachPlayer(player: Player) {
        deferredLicenseManager.attach(player)
        super.attachPlayer(player)
    }

    override fun detachPlayer() {
        deferredLicenseManager.detach()
        super.detachPlayer()
    }

    /**
     * Bitmovin Analytics
     *
     * @param bitmovinAnalyticsConfig [BitmovinAnalyticsConfig]
     * @param context [Context]
     */
    @Deprecated(
        "Please use {@link IBitmovinCollector.Factory instead} instead.",
        replaceWith = ReplaceWith(
            expression = "IBitmovinPlayerCollector.Factory.create(context, analyticsConfig)",
            imports = ["com.bitmovin.analytics.bitmovin.player.api.IBitmovinPlayerCollector"],
        ),
    )
    constructor(bitmovinAnalyticsConfig: BitmovinAnalyticsConfig, context: Context) : this(
        ApiV3Utils.extractAnalyticsConfig(bitmovinAnalyticsConfig),
        context,
    ) {
        this.setDeprecatedBitmovinAnalyticsConfig(bitmovinAnalyticsConfig)
    }

    override fun createAdapter(
        player: Player,
        analytics: BitmovinAnalytics,
    ): PlayerAdapter {
        val licenseKeyProvider = deferredLicenseManager.licenseKeyProvider
        val featureFactory: FeatureFactory = BitmovinFeatureFactory(
            analytics,
            player,
            licenseKeyProvider,
        )
        val userAgentProvider = UserAgentProvider(
            Util.getApplicationInfoOrNull(analytics.context),
            Util.getPackageInfoOrNull(analytics.context),
            SystemInformationProvider.getProperty("http.agent"),
        )
        val eventDataFactory = EventDataFactory(
            config,
            userIdProvider,
            userAgentProvider,
            licenseKeyProvider,
        )
        val deviceInformationProvider = DeviceInformationProvider(analytics.context)
        val playerLicenseProvider = PlayerLicenseProvider(analytics.context)
        val playerContext = BitmovinPlayerContext(player)
        val handler = Handler(analytics.context.mainLooper)
        val stateMachine = PlayerStateMachine.Factory.create(analytics, playerContext, handler)
        val playbackQualityProvider = PlaybackQualityProvider(player)
        return BitmovinSdkAdapter(
            player,
            config,
            stateMachine,
            featureFactory,
            eventDataFactory,
            deviceInformationProvider,
            playerLicenseProvider,
            playbackQualityProvider,
            metadataProvider,
        )
    }

    @Deprecated(
        "Use setSourceMetadata instead",
        ReplaceWith("setSourceMetadata(sourceMetadata, playerSource)"),
    )
    override fun addSourceMetadata(playerSource: Source, sourceMetadata: SourceMetadata) {
        this.setSourceMetadata(playerSource, sourceMetadata)
    }

    override fun setSourceMetadata(playerSource: Source, sourceMetadata: SourceMetadata) {
        metadataProvider.setSourceMetadata(playerSource, sourceMetadata)
    }

    override fun getSourceMetadata(playerSource: Source): SourceMetadata {
        return metadataProvider.getSourceMetadata(playerSource) ?: SourceMetadata()
    }

    override fun setCustomData(playerSource: Source, customData: CustomData) {
        // we cannot put this logic into the adapter since the adapter is created on attaching
        // and this method might be called earlier
        val newActiveCustomData =
            ApiV3Utils.mergeCustomData(customData, metadataProvider.defaultMetadata.customData)
        val activeCustomDataChanged = analytics.activeCustomData != newActiveCustomData

        if (playerSource.isActive && activeCustomDataChanged) {
            analytics.closeCurrentSampleForCustomDataChangeIfNeeded()
        }

        val sourceMetadata = metadataProvider.getSourceMetadata(playerSource)
        if (sourceMetadata != null) {
            metadataProvider.setSourceMetadata(
                playerSource,
                sourceMetadata.copy(customData = customData),
            )
        } else {
            metadataProvider.setSourceMetadata(
                playerSource,
                SourceMetadata(customData = customData),
            )
        }
    }

    override fun getCustomData(playerSource: Source): CustomData {
        val currentSourceMetadata = metadataProvider.getSourceMetadata(playerSource)
        return currentSourceMetadata?.customData ?: CustomData()
    }
}
